### PR TITLE
Add last_in_cidr function

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

**Comment by igalic** [comment]
wouldn't this specific function be better placediin our network module?

**Comment by bastelfreak** [comment]
I'm not sure about this.the network module is pretty old and suited for CentOS 6 and older, it does not support systemd-networkd. I'm not sure if it still provides much benefits. it has a bunch of broken tests. in my case I only need the functions and not the types/providers from our network module and I assume other people think the same way.

**Review by alexjfisher — CHANGES_REQUESTED** [review]
Could you add some tests?

File: REFERENCE.md [hunk 1]
```diff
@@ -92,7 +94,43 @@ Data type: `Any`
...
+
+##### calling the function
+
+```puppet
+extlib::cidr_to_broadcast('127.0.0.1/8')
+```
+
+#### `extlib::cidr_to_broadcast(Variant[Stdlib::IP::Address::V4::CIDR,Stdlib::IP::Address::V6::CIDR] $ip)`
+
+this is based on https://github.com/Yelp/puppet-netstdlib/blob/5a62ea51c1371b797e905afe6aa2d519db92e062/lib/puppet/parser/functions/cidr_to_broadcast.rb#L1
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply REFERENCE.md:114]: fine with me. I just wanted to give Yelp some credit for the initial work

File: lib/puppet/functions/extlib/cidr_to_broadcast.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# this is based on https://github.com/Yelp/puppet-netstdlib/blob/5a62ea51c1371b797e905afe6aa2d519db92e062/lib/puppet/parser/functions/cidr_to_broadcast.rb#L1
+# Imported by Tim 'bastelfreak' Meusel into voxpupuli/extlib because Yelp/netstdlib got abandoned
+#
+# @summary Converts an CIDR address of the form 192.168.0.1/24 into its broadcast address
+#
+Puppet::Functions.create_function(:'extlib::cidr_to_broadcast') do
+  # @param ip IPv6 or legacy IPv4 address in CIDR notation
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply lib/puppet/functions/extlib/cidr_to_broadcast.rb:7]: `useful` might be the wrong term here. I had this RFC in mind: https://tools.ietf.org/html/draft-howard-sunset4-v4historic-00
but I can remove it if you like :D

**Comment by bastelfreak** [comment]
@alexjfisher I had the tests from the beginning but forgot to commit them :( added them now and addressesed your comments.

File: lib/puppet/functions/extlib/cidr_to_broadcast.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Imported by Tim 'bastelfreak' Meusel into voxpupuli/extlib because Yelp/netstdlib got abandoned
+#
+# @summary Converts an CIDR address of the form 192.168.0.1/24 into its broadcast address
+#
+Puppet::Functions.create_function(:'extlib::cidr_to_broadcast') do
+  # @param ip IPv6 or IPv4 address in CIDR notation
```
Comment (ekohl) [inline lib/puppet/functions/extlib/cidr_to_broadcast.rb:6]: IPv6 doesn't have the concept of broadcast so I would reject it.

File: REFERENCE.md [hunk 4]
```diff
@@ -142,7 +178,7 @@ Data type: `Optional[String]`
 
 The path to an .erb or .epp template file or `undef`.
 
-### extlib::dir_split
+### `extlib::dir_split`
```
Comment (ekohl) [inline REFERENCE.md:181]: Perhaps we should first regen the README in a separate PR/commit

Comment (bastelfreak) [reply REFERENCE.md:181]: done in https://github.com/voxpupuli/puppet-extlib/pull/169

File: lib/puppet/functions/extlib/cidr_to_broadcast.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexjfisher) [reply lib/puppet/functions/extlib/cidr_to_broadcast.rb]: yeah, it's a bit opinionated.

File: spec/functions/extlib/cidr_to_broadcast.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  context 'when called with an IP Address that is not in the CIDR notation' do
+    it { is_expected.to run.with_params('fe80::800:27ff:fe00:0').and_raise_error(ArgumentError) }
+  end
+
+  context 'when called with an IPv4 CIDR' do
+    it { is_expected.to run.with_params('127.0.0.1/8').and_return('127.255.255.255') }
+  end
+
+  context 'when called with an IPv6 CIDR' do
+    it { is_expected.to run.with_params('fe80::800:27ff:fe00:0/64').and_return('fe80::ffff:ffff:ffff:ffff') }
```
Comment (alexjfisher) [inline spec/functions/extlib/cidr_to_broadcast.rb:33]: Reading @ekohl 's comment, perhaps it shouldn't though.

File: lib/puppet/functions/extlib/cidr_to_broadcast.rb (hunk 3 above)
Comment (bastelfreak) [reply lib/puppet/functions/extlib/cidr_to_broadcast.rb]: dropped IPv6

File: spec/functions/extlib/cidr_to_broadcast.rb (hunk 5 above)
Comment (bastelfreak) [reply spec/functions/extlib/cidr_to_broadcast.rb]: now it raises an argument error

**Review by alexjfisher — APPROVED** [review]
Looks good to me.

**Review by ekohl — APPROVED** [review]

**Comment by kenyon** [comment]
Since this is really just getting the last address in the range, maybe it would be more useful to call it `last_in_cidr` or something, and keep the IPv6 functionality? Alternatively, it would be good to have a function that gets an address of a subnet given an index (like the [`ipaddr` Jinja filter in Ansible](https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters_ipaddr.html#basic-queries)), then you could just get the `-1` index for the last address.

**Comment by logicminds** [comment]
Can you rebase and think about renaming function commented by @kenyon

**Comment by bastelfreak** [comment]
@logicminds can you take a look please?

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Imported by Tim 'bastelfreak' Meusel into voxpupuli/extlib because Yelp/netstdlib got abandoned
+#
+# @summary Converts an IPv4 CIDR address of the form 192.168.0.1/24 into its broadcast address
+# @note IPv6 supports only multicast and unicast, so this function only supports IPv4
```
Comment (b4ldr) [inline lib/puppet/functions/extlib/last_in_cidr.rb:4]: now the function doesn't mention `broadcast` this restriction can be removed

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+Puppet::Functions.create_function(:'extlib::last_in_cidr') do
+  # @param ip IPv4 address in CIDR notation
+  # @return IPv4 broadcast address
+  # @example calling the function
+  #   extlib::cidr_to_broadcast('127.0.0.1/8')
+  dispatch :last_in_cidr do
+    param 'Stdlib::IP::Address::V4::CIDR', :ip
```
Comment (b4ldr) [inline lib/puppet/functions/extlib/last_in_cidr.rb:11]: which means this becomes Variant[v4,v6]

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    return_type 'Stdlib::IP::Address::V4::Nosubnet'
```
Comment (b4ldr) [inline lib/puppet/functions/extlib/last_in_cidr.rb:12]: ditto

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# Imported by Tim 'bastelfreak' Meusel into voxpupuli/extlib because Yelp/netstdlib got abandoned
+#
+# @summary Converts an IPv4 CIDR address of the form 192.168.0.1/24 into its last address
+# @note IPv6 supports only multicast and unicast, so this function only supports IPv4
+Puppet::Functions.create_function(:'extlib::last_in_cidr') do
+  # @param ip IPv4 address in CIDR notation
+  # @return IPv4 broadcast address
```
Comment (kenyon) [inline lib/puppet/functions/extlib/last_in_cidr.rb:7]:
```suggestion
  # @return last address in the network
```

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  # @example calling the function
+  #   extlib::cidr_to_broadcast('127.0.0.1/8')
```
Comment (kenyon) [inline lib/puppet/functions/extlib/last_in_cidr.rb:9]:
```suggestion
  #   extlib::last_in_cidr('127.0.0.1/8')
```

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+# Imported by Tim 'bastelfreak' Meusel into voxpupuli/extlib because Yelp/netstdlib got abandoned
+#
+# @summary Converts an IPv4 CIDR address of the form 192.168.0.1/24 into its last address
+# @note IPv6 supports only multicast and unicast, so this function only supports IPv4
```
Comment (kenyon) [inline lib/puppet/functions/extlib/last_in_cidr.rb:4]:
```suggestion
```

File: spec/functions/extlib/last_in_cidr.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  context 'when called with an IP Address that is not in the CIDR notation' do
+    it { is_expected.to run.with_params('fe80::800:27ff:fe00:0').and_raise_error(ArgumentError) }
+  end
+
+  context 'when called with an IPv4 CIDR' do
+    it { is_expected.to run.with_params('127.0.0.1/8').and_return('127.255.255.255') }
+  end
+
+  context 'when called with an IPv6 CIDR' do
+    it { is_expected.to run.with_params('fe80::800:27ff:fe00:0/64').and_raise_error(ArgumentError) }
```
Comment (kenyon) [inline spec/functions/extlib/last_in_cidr.rb:33]:
```suggestion
    it { is_expected.to run.with_params('2001:db8::1/64').and_return('2001:db8::ffff:ffff:ffff:ffff') }
```

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+Puppet::Functions.create_function(:'extlib::last_in_cidr') do
+  # @param ip IPv4 address in CIDR notation
```
Comment (kenyon) [inline lib/puppet/functions/extlib/last_in_cidr.rb:6]:
```suggestion
  # @param ip IP address in CIDR notation
```

File: lib/puppet/functions/extlib/last_in_cidr.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# Imported by Tim 'bastelfreak' Meusel into voxpupuli/extlib because Yelp/netstdlib got abandoned
+#
+# @summary Converts an IPv4 CIDR address of the form 192.168.0.1/24 into its last address
```
Comment (kenyon) [inline lib/puppet/functions/extlib/last_in_cidr.rb:3]:
```suggestion
# @summary Converts an IPv4 or IPv6 CIDR address of the form 192.0.2.1/24 or 2001:db8::1/64 into the last address in the network
```

**Review by kenyon — CHANGES_REQUESTED** [review]
Also regenerate `REFERENCE.md`.

**Review by kenyon — APPROVED** [review]

**Pull request merged**